src/Database/LSMTree/Internal/Run/Index/Compact/Construction.hs

```diff
@@ -249,45 +250,54 @@ yield :: MCompactIndex s -> ST s (Maybe Chunk)
 yield MCompactIndex{..} = do
     pageNo <- readSTRef mciCurrentPageNumber
     if pageNo `mod` mciMaxChunkSize == 0 then do -- The current chunk is full
-      cPrimary <- VU.unsafeFreeze =<< readSTRef mciPrimary
-      (writeSTRef mciPrimary $!) =<< VUM.new mciMaxChunkSize
+      cPrimary <- VU.unsafeFreeze . NE.head =<< readSTRef mciPrimary
+      modifySTRef' mciPrimary . NE.cons =<< VUM.new mciMaxChunkSize
       modifySTRef' mciClashes . NE.cons =<< VUM.new mciMaxChunkSize
       modifySTRef' mciLargerThanPage . NE.cons =<< VUM.new mciMaxChunkSize
       pure $ Just (Chunk cPrimary)
     else -- the current chunk is not yet full
       pure Nothing
 
+newtype PrimaryArray = PrimaryArray (VU.Vector Word32)
+
```

Comment:
Related to my earlier comment on `unsafeEnd`  -- this could be moved together with `fromChunks`